### PR TITLE
KAFKA-5038: Catch exception

### PR DESCRIPTION
**Pull request by enothereska**

### CODE REVIEW

**Comment by enothereska** [comment]
@dguy @mjsax for a review. Thanks.

**Comment by enothereska** [comment]
Streams tests also pass on branch builder: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/256/

**Review by mjsax — APPROVED** [review]
LGTM

**Review by dguy — APPROVED** [review]
Thanks @eno, LGTM

**Comment by dguy** [comment]
@ijuma can you please merge this?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 1]
```diff
@@ -95,11 +95,18 @@ public File globalStateDir() {
...
      */
     public boolean lock(final TaskId taskId, int retry) throws IOException {
+        final File lockFile;
         // we already have the lock so bail out here
         if (locks.containsKey(taskId)) {
             return true;
         }
-        final File lockFile = new File(directoryForTask(taskId), LOCK_FILE_NAME);
+        try {
+            lockFile = new File(directoryForTask(taskId), LOCK_FILE_NAME);
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:104]: Is it OK that we call `directoryForTask` with no protection in `ProcessorStateManager` before we call `lock`?

```
        this.baseDir  = stateDirectory.directoryForTask(taskId);
        this.partitionForTopic = new HashMap<>();
        for (TopicPartition source : sources) {
            this.partitionForTopic.put(source.topic(), source);
        }
        this.stores = new LinkedHashMap<>();
        this.globalStores = new HashMap<>();
        this.offsetLimits = new HashMap<>();
        this.restoredOffsets = new HashMap<>();
        this.isStandby = isStandby;
        this.restoreCallbacks = isStandby ? new HashMap<String, StateRestoreCallback>() : null;
        this.storeToChangelogTopic = storeToChangelogTopic;

        this.logPrefix = String.format("task [%s]", taskId);

        if (!stateDirectory.lock(taskId, 5)) {
            throw new LockException(String.format("%s Failed to lock the state directory: %s", logPrefix, baseDir.getCanonicalPath()));
        }
```

Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:104]: Good catch @ijuma . In this case crashing is probably the right thing to do, however I've now caught the exception and propagated further so at least we know why we're crashing.

**Comment by enothereska** [comment]
System test passed https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/257/

**Comment by enothereska** [comment]
@dguy mind having another look at recent change? Thanks.

**Review by dguy — APPROVED** [review]
LGTM

**Review by ijuma — APPROVED** [review]
LGTM, merged to 0.10.2. Please file another PR for trunk.

**Pull request closed without merging**